### DISCUSSION TITLE
Multiple instances of the same project

### DISCUSSION THREAD

**Discussion by antunesl** [microsoft/aspire#1022] [Q&A] [answered]
Hi!

In my solution I have a project that is used as a "Shell" for all modules (each module has a collection of endpoints, like a microservice), the modules to load for an instance of this are passed by configuration. So the product can behave like a monolith or scale by selecting modules to deploy in separate when needed.

The goal is to be able to have different deployment configurations based on the need of it and to grow in a easy and mor cost effective way, and then scale as needed. For example, in local dev I deploy all  modules in the same running webapp, but by configuration I can run a different webapp for each module.

I'm in a process of exploring Aspire, it seems a perfect tool for this kind of distributed projects, but I'm facing a issue here when trying to launch one instance per module. The pattern I'm using here (don't know if it is the best) is to declare the ProjectResources and setting the launchProfile for a specific value for each. This way I can use Env Variables on the specific launchprofile to tell what modules to load for each instance. This is the way I imagine to "pass data" from the Host to the running instances.

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var apiGateway = builder.AddProject<Projects.Lynx_Backend_APIGateway>("apigateway");
var db = builder
    .AddSqlServerContainer("dbServer")
    .AddDatabase("db");

// Projects.Lynx_Backend_Host is the host project for every module

var moduleA = builder
    .AddProject<Projects.Lynx_Backend_Host>("Core")
    .WithLaunchProfile("core")
    .WithReference(db);

var moduleB = builder
    .AddProject<Projects.Lynx_Backend_Host>("HR")
    .WithLaunchProfile("hr")
    .WithReference(db);

apiGateway.WithReference(moduleA);
apiGateway.WithReference(moduleB);

builder.Build().Run();
```

On running this for 2 instances (each one will load a module and run), I get the following error:
```
 Aspire.Hosting.Dashboard.ProjectViewModelCache[0]
      Task to write view model change terminated for resource type: ProjectViewModel
      System.InvalidOperationException: Sequence contains more than one matching element
         at System.Linq.ThrowHelper.ThrowMoreThanOneMatchException()
         at System.Linq.Enumerable.TryGetSingle[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
         at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
         at Aspire.Hosting.ApplicationModel.ProjectResourceExtensions.TryGetProjectWithPath(DistributedApplicationModel model, String path, ProjectResource& projectResource) in /_/src/Aspire.Hosting/ApplicationModel/ProjectResourceExtensions.cs:line 32
         at Aspire.Hosting.Dashboard.ViewModelCache`2.<>c__DisplayClass12_0.<FillEndpoints>b__1(Endpoint ep) in /_/src/Aspire.Hosting/Dashboard/ViewModelCache.cs:line 186
         at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
         at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
         at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
         at Aspire.Hosting.Dashboard.ViewModelCache`2.FillEndpoints(DistributedApplicationModel applicationModel, IEnumerable`1 services, IEnumerable`1 endpoints, CustomResource resource, ResourceViewModel resourceViewModel) in /_/src/Aspire.Hosting/Dashboard/ViewModelCache.cs:line 174
         at Aspire.Hosting.Dashboard.ProjectViewModelCache.ConvertToViewModel(DistributedApplicationModel applicationModel, IEnumerable`1 services, IEnumerable`1 endpoints, Executable executable, List`1 additionalEnvVars) in /_/src/Aspire.Hosting/Dashboard/ProjectViewModelCache.cs:line 40
         at Aspire.Hosting.Dashboard.ViewModelCache`2.ViewModelGeneratingEnumerator.ComputeResult(ObjectChangeType objectChangeType, TResource resource, List`1 additionalEnvVars) in /_/src/Aspire.Hosting/Dashboard/ViewModelCache.cs:line 462
         at Aspire.Hosting.Dashboard.ViewModelCache`2.ViewModelGeneratingEnumerator.MoveNextAsync() in /_/src/Aspire.Hosting/Dashboard/ViewModelCache.cs:line 352
         at Aspire.Hosting.Dashboard.ViewModelCache`2.<>c__DisplayClass7_0.<<-ctor>b__0>d.MoveNext() in /_/src/Aspire.Hosting/Dashboard/ViewModelCache.cs:line 59

```

Checking the source code I can see that this maybe not going to work, because it checks for projects with the "same path", and in my case they all are the same project running - just with different context.

```csharp
    /// <summary>
    /// Tries to get the project resource with the specified path from the distributed application model.
    /// </summary>
    /// <param name="model">The distributed application model.</param>
    /// <param name="path">The path of the project resource.</param>
    /// <param name="projectResource">When this method returns, contains the project resource with the specified path, if it is found; otherwise, null.</param>
    /// <returns><see langword="true"/> if the project resource with the specified path is found; otherwise, <see langword="false"/>.</returns>
    public static bool TryGetProjectWithPath(this DistributedApplicationModel model, string path, [NotNullWhen(true)] out ProjectResource? projectResource)
    {
        projectResource = model.GetProjectResources().SingleOrDefault(p => p.Annotations.OfType<IServiceMetadata>().FirstOrDefault()?.ProjectPath == path);

        return projectResource is not null;
    }
```

Is this pattern some thing supported right now? Or in the future?
Maybe I'm thinking something wrong in my approach to this, does anyone with the same type of solution architecture that has solved this?

Thanks.

**Comment by davidfowl** [comment, marked as answer]
That approach looks reasonable to be. Given that you *do* want to model launching the same code as different services. We just have a bug in this `TryGetProjectWithPath` implementation that assumes there would only be one project with this path. We should be able to solve this.

**Reply by antunesl** [reply]
Thanks for the feedback, hope that's a thing you guys can fix for the next release.

Do you think this can be treat in different/better way using Aspire (other than using launch profiles for "configuration")?
I would prefer to have the configuration on the AppHost (aspire) and then pass that to the project resources, but I don't see a way to do that.

**Reply by davidfowl** [reply]
What does “pass configuration” look like?

**Reply by antunesl** [reply]
For me it's a json, where I can declare all modules being loaded by that "project resource". Each one can run and load 1 or more modules (a module is just a class lib).
Without aspire, I run the Host project that reads a json and do the appropriate, but with Aspire the ideal use case was to be able to "pass some context" to each project resource. Maybe some custom env variable with the json is generic enough to pass some data, this way the configuration stays on the host for flexibility (this is important for deployment).

```csharp
var moduleA = builder
    .AddProject<Projects.Lynx_Backend_Host>("Core")
    .WithEnvironmentVariable("HOST_CONFIG", "{...}")
    .WithReference(db);
```

**Reply by antunesl** [reply]
> We should be able to solve this.

Should I open a issue with this?

**Reply by davidfowl** [reply]
Yep.

**Reply by siby-george** [reply]
Linking the issue #1072 with fix

**Comment by raffaeler** [comment]
How can you show a custom name in the dashboard for each instance of the same project?

**Reply by davidfowl** [reply]
What is a custom name? The resource name isn't the same as the project name.

**Reply by raffaeler** [reply]
Having multiple instances of the same project, it's more convenient to assign a friendly name to the instance rather then seeing multiple times the same project. This way, you can distinguish the instances.

**Reply by davidfowl** [reply]
The names have to be distinguished because the resource name cannot be the same.

**Reply by raffaeler** [reply]
My fault, I gave long names and appeared equal on the dashboard, therefore I was asking how to change the `source` instead.
Thanks

**Comment by NakibOrbitax** [comment]
@davidfowl, Hi, I can run multiple instances with different environment variables; however, after running, my project reads some license files, and it causes exceptions because multiple processes(instances) are trying to read the file. 
If the publishing were done in different directories for different instances, then it would be solved. 

Is there any other workaround?
